### PR TITLE
[Edge] Add tcp implementation @open sesame 06/15 18:38

### PR DESCRIPTION
**Pull request by gichan-jang**
- Add NNStreamer-edge TCP implementation.

Signed-off-by: gichan <gichan2.jang@samsung.com>

**Self evaluation:**
1. Build test: [ *]Passed [ ]Failed [ ]Skipped
2. Run test: [ ]Passed [ ]Failed [*]Skipped

### CODE REVIEW

File: gst/nnstreamer/tensor_query/nnstreamer_edge.h [hunk 1]
```diff
@@ -54,6 +54,9 @@ typedef enum {
 
 typedef enum {
   NNS_EDGE_EVENT_UNKNOWN = 0,
+  NNS_EDGE_EVENT_ACCEPATABLE = 1,
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge.h:57]: typo : ACCEPTABLE

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 2]
```diff
@@ -43,37 +203,108 @@ nns_edge_handle_is_valid (nns_edge_h edge_h)
...
 
   if (!id || *id == '\0') {
     nns_edge_loge ("Invalid param, given ID is invalid.");
     return NNS_EDGE_ERROR_INVALID_PARAMETER;
   }
-
   if (!topic || *topic == '\0') {
-    nns_edge_loge ("Invalid param, given topic is invalid.");
-    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+    nns_edge_logw
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:306]: Q: the case if topic is null or empty.
Why did you change the condition about invalid topic string?

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 3]
```diff
@@ -86,31 +317,216 @@ nns_edge_get_handle (const char *id, const char *topic, nns_edge_h * edge_h)
    * 1. consider adding hash table or list to manage edge handles.
    * 2. compare topic and return error if existing topic in handle is different.
    */
-  eh = (nns_edge_handle_s *) malloc (sizeof (nns_edge_handle_s));
+  eh = g_try_new0 (nns_edge_handle_s, 1);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:320]: line 89 and 95.
plz do not change this, we should remove glib dependency later.

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 4]
```diff
@@ -178,33 +1011,74 @@ nns_edge_connect (nns_edge_h edge_h, nns_edge_protocol_e protocol,
...
+  conn_list = g_hash_table_get_keys (_table);
+  while (conn_list) {
+    intptr_t client_id = (intptr_t) conn_list->data;
+    conn_data = _nns_edge_get_conn (client_id);
+
+    conn_data = _nns_edge_get_conn (client_id);
+    if (false == _nns_edge_close_connection (conn_data->sink_conn) ||
+        false == _nns_edge_close_connection (conn_data->src_conn)) {
+      nns_edge_loge ("Failed to close the connection");
+      return NNS_EDGE_ERROR_CONNECTION_FAILURE;
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:1076]: possible mem leak (conn_list).

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 5]
```diff
@@ -329,3 +1215,97 @@ nns_edge_get_topic (nns_edge_h edge_h, char **topic)
...
+    ret_str = g_strdup_printf ("%s%s", eh->caps_str, value);
+    g_free (eh->caps_str);
+    eh->caps_str = ret_str;
+  } else if (0 == g_ascii_strcasecmp (key, "IP")) {
+    g_free (eh->recv_ip);
+    eh->recv_ip = g_strdup (value);
+  } else if (0 == g_ascii_strcasecmp (key, "PORT")) {
+    eh->recv_port = g_ascii_strtoll (value, NULL, 10);
+  } else {
+    nns_edge_logw ("Failed to set edgo info. Unknown key: %s", key);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:1246]: typo : edgo > edge

File: gst/nnstreamer/tensor_query/nnstreamer_edge.h (hunk 1 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/nnstreamer_edge.h]: Thanks. Updated!

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c (hunk 3 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c]: Changed. Thanks.

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c (hunk 4 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c]: Fixed. Thanks.

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c (hunk 5 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c]: Updated.

File: gst/nnstreamer/meson.build [hunk 6]
```diff
@@ -126,7 +126,7 @@ nnstreamer_deps += nnstreamer_single_dep
 # Build libraries ("both_libraries" are supported from 0.46.)
 nnstreamer_shared = shared_library('nnstreamer',
   nnstreamer_sources,
-  dependencies: nnstreamer_deps,
```
Comment (jaeyun-jung) [inline gst/nnstreamer/meson.build:129]: plz move this to next PR.

Comment (jaeyun-jung) [inline gst/nnstreamer/meson.build:129]: plz move this to next PR.

Comment (gichan-jang) [reply gst/nnstreamer/meson.build]: Moved. Thanks

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 7]
```diff
@@ -13,30 +13,250 @@
...
+  GList *conn_list = NULL;
+  bool ret = true;
 
   if (!NNS_EDGE_MAGIC_IS_VALID (eh)) {
     nns_edge_loge ("Invalid param, given edge handle is invalid.");
     return false;
   }
 
-  /** @todo check connection status */
+  conn_list = g_hash_table_get_values (_table);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:200]: get list w/lock

Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:200]: get list w/lock

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 8]
```diff
@@ -91,15 +452,64 @@ nns_edge_release_handle (nns_edge_h edge_h)
...
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+
   if (eh->event_cb) {
     /** @todo send new event (release handle) */
     eh->event_cb = NULL;
   }
 
   eh->magic = NNS_EDGE_MAGIC_DEAD;
+  nns_edge_disconnect (edge_h);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:509]: disconnect before updating magic

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c (hunk 7 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:200]: Updated.

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c (hunk 8 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c]: Updated.

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 9]
```diff
@@ -131,16 +546,435 @@ nns_edge_set_event_callback (nns_edge_h edge_h, nns_edge_event_cb cb,
...
+   /** Create message receving thread */
+  pthread_attr_init (&attr);
+  pthread_attr_setdetachstate (&attr, PTHREAD_CREATE_DETACHED);
+
+  thread_data =
+      (nns_edge_thread_data_s *) malloc (sizeof (nns_edge_thread_data_s));
+  if (!thread_data) {
+    _nns_edge_close_connection (conn);
+    nns_edge_loge ("Failed to allocate query thread data.");
+  }
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:869]: need return here

Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:869]: need return here

Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:869]: and plz move 861~862 (thread attr) after this line

Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:869]: and plz move 861~862 (thread attr) after this line

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 10]
```diff
@@ -131,16 +546,435 @@ nns_edge_set_event_callback (nns_edge_h edge_h, nns_edge_event_cb cb,
...
+  nns_edge_handle_s *eh = (nns_edge_handle_s *) edge_h;
+  nns_edge_query_cmd_buf_s cmd_buf;
+  nns_edge_conn_s *conn = NULL;
+  nns_edge_event_h event_h = NULL;
+  int64_t client_id;
+  nns_edge_conn_data_s *conn_data;
+  conn = (nns_edge_conn_s *) malloc (sizeof (nns_edge_conn_s));
+  if (!conn) {
+    nns_edge_loge ("Failed to allocate client data.");
+    return NNS_EDGE_ERROR_CONNECTION_FAILURE;
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:906]: error code out-of-memory?

Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:906]: error code out-of-memory?

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 11]
```diff
@@ -325,3 +1216,102 @@ nns_edge_get_topic (nns_edge_h edge_h, char **topic)
...
+
+  eh = (nns_edge_handle_s *) edge_h;
+
+  if (!NNS_EDGE_MAGIC_IS_VALID (eh)) {
+    nns_edge_loge ("Invalid param, given edge handle is invalid.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+
+  if (0 == g_ascii_strcasecmp (key, "CAPS")) {
+    ret_str = g_strdup_printf ("%s%s", eh->caps_str, value);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:1238]: Q: replace old value or append new string
IMO replacing caps string is better (user handles the capability of edge handle)

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 12]
```diff
@@ -325,3 +1216,102 @@ nns_edge_get_topic (nns_edge_h edge_h, char **topic)
...
+  char *ret_str = NULL;
+
+  eh = (nns_edge_handle_s *) edge_h;
+
+  if (!NNS_EDGE_MAGIC_IS_VALID (eh)) {
+    nns_edge_loge ("Invalid param, given edge handle is invalid.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+
+  if (0 == g_ascii_strcasecmp (key, "CAPS")) {
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:1237]: Q: is it necessary to handle key-value set as json or table?

Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:1237]: Q: is it necessary to handle key-value set as json or table?

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c (hunk 11 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c]: Marked as todo item. Thanks.

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c (hunk 12 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:1242]: Marked as todo item. Thanks.

**Review by jaeyun-jung — APPROVED** [review]

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 13]
```diff
@@ -46,23 +254,164 @@ nns_edge_get_handle (const char *id, const char *topic, nns_edge_h * edge_h)
...
 
-  memset (eh, 0, sizeof (nns_edge_handle_s));
   eh->magic = NNS_EDGE_MAGIC;
   eh->id = g_strdup (id);
   eh->topic = g_strdup (topic);
-  eh->protocol = NNS_EDGE_PROTOCOL_MAX;
+  eh->protocol = NNS_EDGE_PROTOCOL_TCP;
+  eh->is_server = true;
+  eh->recv_ip = g_strdup ("localhost");
+  eh->recv_ip = 0;
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:265]: maybe port

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 14]
```diff
@@ -72,16 +421,69 @@ nns_edge_release_handle (nns_edge_h edge_h)
...
+          &saddr)) {
+    nns_edge_loge ("Failed to get socket address");
+    ret = -EADDRNOTAVAIL;
+    goto error;
+  }
+  if (!g_socket_listener_add_address (eh->listener, saddr,
+          G_SOCKET_TYPE_STREAM, G_SOCKET_PROTOCOL_TCP, NULL, NULL, &err)) {
+    nns_edge_loge ("Failed to add address: %s", err->message);
+    g_clear_error (&err);
+    ret = -EADDRNOTAVAIL;
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:443]: let's define edge error code for this. (Or use NNS_EDGE_ERROR_CONNECTION_FAILURE)

Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:443]: Updated to use `NNS_EDGE_ERROR_CONNECTION_FAILURE`. Thanks.

**Review by again4you — APPROVED** [review]
LGTM!
